In [1]:
import pandas as pd
import numpy as np
from pybaseball import pitching_stats

# Vertical Release Point Analysis

In [2]:
RP_comparison = pd.read_csv(r"C:\Users\Owner\OneDrive\Desktop\exploratory data files\Pitch Analysis\Release Points.csv")[["player_name","RP_z", "RP_x","ba", "iso", "babip", "slg", "woba", "hits", "launch_speed", "launch_angle", "spin_rate", "velocity", "whiffs"]]
RP_comparison['Type'] = np.where(RP_comparison['RP_z'] < 4.7, "Submarine", "Normal")
RP_comparison.columns

Index(['player_name', 'RP_z', 'RP_x', 'ba', 'iso', 'babip', 'slg', 'woba',
       'hits', 'launch_speed', 'launch_angle', 'spin_rate', 'velocity',
       'whiffs', 'Type'],
      dtype='object')

In [3]:
RP_comparison['vert_quartile'] = np.where((RP_comparison['RP_z'] < RP_comparison['RP_z'].quantile(.1)), '10% less',
    np.where((RP_comparison['RP_z'] < RP_comparison['RP_z'].quantile(.25)) & (RP_comparison['RP_z'] >= RP_comparison['RP_z'].quantile(.1)),'10-25%',
    np.where((RP_comparison['RP_z'] < RP_comparison['RP_z'].quantile(.5)) & (RP_comparison['RP_z'] >= RP_comparison['RP_z'].quantile(.25)), '25-50%',
    np.where((RP_comparison['RP_z'] < RP_comparison['RP_z'].quantile(.75)) & (RP_comparison['RP_z'] >= RP_comparison['RP_z'].quantile(.5)), '50-75%',
    np.where((RP_comparison['RP_z'] < RP_comparison['RP_z'].quantile(.90)) & (RP_comparison['RP_z'] >= RP_comparison['RP_z'].quantile(.75)), '75-90%', '90% >')))))

In [4]:
RP_comparison['horiz_quartile'] = np.where((RP_comparison['RP_x'] < RP_comparison['RP_x'].quantile(.1)), '10%>',
    np.where((RP_comparison['RP_x'] < RP_comparison['RP_x'].quantile(.25)) & (RP_comparison['RP_x'] >= RP_comparison['RP_x'].quantile(.1)),'10-25%',
    np.where((RP_comparison['RP_x'] < RP_comparison['RP_x'].quantile(.5)) & (RP_comparison['RP_x'] >= RP_comparison['RP_x'].quantile(.25)), '25-50%',
    np.where((RP_comparison['RP_x'] < RP_comparison['RP_x'].quantile(.75)) & (RP_comparison['RP_x'] >= RP_comparison['RP_x'].quantile(.5)), '50-75%',
    np.where((RP_comparison['RP_x'] < RP_comparison['RP_x'].quantile(.90)) & (RP_comparison['RP_x'] >= RP_comparison['RP_x'].quantile(.75)), '75-90%', '90%>')))))

In [5]:
pitchers = pitching_stats(2022, qual = 5)[["Name","WAR", "ERA", "K/9", "BB/9", "WHIP", "FIP", 'SIERA', 'Hard%', 'Soft%']]

In [6]:
pitchers[['first', 'last']] = pitchers['Name'].str.split(' ', n= 1, expand= True)
pitchers['player_name'] = pitchers['last'] + ', ' + pitchers['first']
pitchers.drop(columns = ['Name', 'first', 'last'], inplace = True)
pitchers

WAR   ERA    K/9  BB/9  WHIP   FIP  SIERA  Hard%  Soft%  \
162  4.8  2.99  10.37  1.53  1.24  2.05   3.00  0.312  0.145   
58   4.7  2.19   8.00  2.24  0.98  2.85   3.50  0.268  0.211   
147  4.7  2.89  11.48  2.63  1.05  2.34   2.96  0.284  0.182   
105  4.4  2.60   8.24  1.55  1.05  2.49   3.30  0.246  0.205   
208  4.4  3.25   9.93  1.26  0.96  2.76   2.89  0.307  0.169   
..   ...   ...    ...   ...   ...   ...    ...    ...    ...   
465 -0.8  5.03   9.53  5.29  1.56  7.35   4.51  0.474  0.124   
577 -0.8  6.41   5.52  4.92  1.69  6.43   5.87  0.359  0.146   
620 -0.8  7.50   7.00  3.00  1.39  7.61   5.01  0.241  0.172   
425 -0.9  4.61   5.60  2.96  1.54  6.30   4.10  0.274  0.116   
571 -1.1  6.33   8.79  3.39  1.46  6.73   4.40  0.313  0.184   

            player_name  
162      Gausman, Kevin  
58     Alcantara, Sandy  
147       Rodon, Carlos  
105          Fried, Max  
208         Nola, Aaron  
..                  ...  
465       Swarmer, Matt  
577         Oller, Adam  
620        Romo, Sergio  
425          Smith, Joe  
571  Hernandez, Elieser  

[666 rows x 10 columns]

In [7]:
RP_stats = pd.merge(RP_comparison, pitchers, left_on = "player_name", right_on = "player_name")

In [8]:
RP_stats

player_name      RP_z      RP_x     ba    iso  babip    slg   woba  \
0         Abbott, Cory  5.667924 -2.599139  0.207  0.276  0.200  0.483  0.347   
1        Abreu, Albert  5.709502 -1.864231  0.239  0.116  0.293  0.355  0.320   
2         Abreu, Bryan  5.770309 -1.985175  0.218  0.058  0.359  0.276  0.273   
3     Acevedo, Domingo  5.475703 -1.997510  0.194  0.183  0.207  0.377  0.270   
4          Adam, Jason  5.894411 -1.692083  0.128  0.071  0.182  0.199  0.185   
..                 ...       ...       ...    ...    ...    ...    ...    ...   
576        Young, Alex  5.884083  2.144250  0.222  0.148  0.300  0.370  0.295   
577       Young, Danny  5.163651  3.847778  0.367  0.167  0.435  0.533  0.408   
578       Zerpa, Angel  5.528742  2.333333  0.220  0.146  0.194  0.366  0.282   
579        Zeuch, T.J.  5.561834 -1.523136  0.436  0.333  0.419  0.769  0.526   
580  Zimmermann, Bruce  6.116975  0.963083  0.322  0.279  0.330  0.601  0.403   

     hits  launch_speed  ...  horiz_quartile  WAR    ERA    K/9  BB/9  WHIP  \
0      18          89.5  ...            10%> -0.3   5.16   8.34  4.76  1.32   
1      33          86.6  ...          25-50% -0.1   3.32   9.00  5.21  1.47   
2      34          86.8  ...          25-50%  0.9   2.28  13.29  3.74  1.22   
3      37          85.9  ...          25-50% -0.1   2.93   7.48  1.95  0.90   
4      20          85.7  ...          25-50%  1.6   1.07  11.01  1.95  0.63   
..    ...           ...  ...             ...  ...    ...    ...   ...   ...   
576     6          89.0  ...          75-90%  0.2   2.08   7.27  3.12  1.50   
577    11          84.1  ...            90%>  0.0   4.26   8.53  2.84  2.05   
578     9          91.5  ...            90%> -0.1   1.64   2.45  2.45  1.09   
579    17          91.6  ...          50-75% -0.3  13.50   5.63  3.38  2.50   
580    89          90.4  ...          75-90% -0.4   5.85   6.12  1.46  1.48   

       FIP  SIERA  Hard%  Soft%  
0     6.56   5.03  0.227  0.136  
1     4.80   4.15  0.234  0.234  
2     1.98   2.51  0.289  0.186  
3     4.14   3.50  0.255  0.183  
4     2.13   2.23  0.219  0.190  
..     ...    ...    ...    ...  
576   2.54   3.90  0.241  0.138  
577   4.22   3.70  0.292  0.208  
578   5.75   5.49  0.474  0.105  
579  10.24   5.04  0.371  0.143  
580   5.79   4.41  0.312  0.128  

[581 rows x 26 columns]

In [9]:
RP_stats.groupby('Type').mean().T

Type               Normal    Submarine
RP_z             5.824702     3.873140
RP_x            -0.754024    -2.183640
ba               0.247208     0.231583
iso              0.155215     0.137333
babip            0.294756     0.278542
slg              0.402424     0.368875
woba             0.318034     0.301125
hits            45.070018    28.916667
launch_speed    88.343806    86.333333
launch_angle    12.853680    10.075000
spin_rate     2244.919210  2254.750000
velocity        89.036984    86.216667
whiffs          97.831239    62.333333
WAR              0.524237     0.245833
ERA              4.459533     3.951667
K/9              8.431616     8.380417
BB/9             3.581562     3.532083
WHIP             1.357828     1.265000
FIP              4.222244     4.005417
SIERA            4.009174     3.723333
Hard%            0.305551     0.274208
Soft%            0.163783     0.202417

## Submarine generally better pitchers but there are less of them so it might be sample size issue
### They also have higher spin rate but a lower velocity
### Normal pitchers generate more whiffs
### How Submarine pitchers pitch however from my understanding have a longevity issue

## ERA, WAR, FIP, SIERA, Contact% are also better
## K/9 roughly the same 

In [10]:
RP_stats.groupby('vert_quartile').mean().T

vert_quartile     10% less       10-25%       25-50%       50-75%  \
RP_z              4.567571     5.361259     5.659636     5.954856   
RP_x             -1.528625    -0.954426    -0.705189    -0.942557   
ba                0.245533     0.250241     0.245250     0.245614   
iso               0.154650     0.164908     0.154804     0.154829   
babip             0.296800     0.301138     0.289500     0.292750   
slg               0.400150     0.415126     0.400061     0.400436   
woba              0.318217     0.328034     0.316122     0.315321   
hits             34.483333    37.643678    42.763514    46.114286   
launch_speed     87.310000    87.839080    88.568243    88.348571   
launch_angle     12.625000    12.116092    13.151351    12.625714   
spin_rate      2281.650000  2277.183908  2239.405405  2229.992857   
velocity         87.116667    88.858621    89.326351    89.531429   
whiffs           74.533333    83.666667    92.418919   103.457143   
WAR               0.368333     0.318391     0.481757     0.595000   
ERA               4.909833     4.808161     4.434932     4.222500   
K/9               8.688333     8.931954     8.232027     8.279000   
BB/9              3.744667     3.922989     3.436419     3.487929   
WHIP              1.366167     1.411379     1.333716     1.335500   
FIP               4.343000     4.470000     4.209865     4.152214   
SIERA             3.878833     3.945287     4.029054     4.014286   
Hard%             0.293850     0.304621     0.313311     0.298707   
Soft%             0.183800     0.160908     0.159412     0.165686   

vert_quartile       75-90%        90% >  
RP_z              6.208079     6.538169  
RP_x             -0.464793    -0.352044  
ba                0.249830     0.242793  
iso               0.145773     0.150172  
babip             0.295511     0.293466  
slg               0.395659     0.392931  
woba              0.316045     0.310293  
hits             52.693182    52.275862  
launch_speed     88.196591    88.977586  
launch_angle     12.484091    13.398276  
spin_rate      2229.931818  2235.431034  
velocity         88.503409    89.001724  
whiffs          108.772727   112.120690  
WAR               0.609091     0.687931  
ERA               4.408750     3.972586  
K/9               8.122273     8.741379  
BB/9              3.555795     3.515690  
WHIP              1.367500     1.331207  
FIP               4.119432     3.992586  
SIERA             4.111250     3.903621  
Hard%             0.296614     0.316362  
Soft%             0.166273     0.166155

## Top release - ISO, SLG, wOBA, more whiffs, WAR, ERA, 
## Submarine - hits, launch speed, soft contact
## The rest seem to be not associated with release point

## When all normal rp are looped into each other they are worse than submarine. However, looking at top RP they seem to be better
## Look at people with a vertical RP over 6.0 but specifically 6.208+

# Horizontal RP analysis

In [11]:
RP_stats.groupby('horiz_quartile').mean().T

horiz_quartile         10%>       10-25%       25-50%       50-75%  \
RP_z               5.062277     5.659291     5.793728     6.005190   
RP_x              -3.044648    -2.310246    -1.799720    -0.887811   
ba                 0.247065     0.244472     0.245085     0.246082   
iso                0.167290     0.153393     0.148943     0.152678   
babip              0.292758     0.290831     0.293071     0.296253   
slg                0.414355     0.397831     0.394000     0.398836   
woba               0.320387     0.316888     0.313418     0.315466   
hits              37.435484    44.449438    47.943262    44.239726   
launch_speed      87.503226    88.465169    88.257447    88.721918   
launch_angle      12.466129    13.050562    12.248227    13.527397   
spin_rate       2273.241935  2253.494382  2245.574468  2272.349315   
velocity          88.325806    89.275281    89.564539    89.495890   
whiffs            81.580645    98.134831   103.496454    98.267123   
WAR                0.379032     0.578652     0.572340     0.579452   
ERA                4.485645     4.482921     4.264113     4.378425   
K/9                8.628710     8.376966     8.431986     8.602877   
BB/9               3.241935     3.666742     3.491418     3.597397   
WHIP               1.317903     1.353708     1.329716     1.350822   
FIP                4.308226     4.249551     4.093191     4.124932   
SIERA              3.784194     4.034944     3.972340     3.950616   
Hard%              0.295258     0.299764     0.303660     0.310479   
Soft%              0.172323     0.160921     0.164142     0.165062   

horiz_quartile       75-90%         90%>  
RP_z               5.960568     5.522929  
RP_x               1.649596     2.807428  
ba                 0.251639     0.246767  
iso                0.156795     0.157017  
babip              0.295639     0.295250  
slg                0.408386     0.403767  
woba               0.322373     0.321633  
hits              47.457831    39.383333  
launch_speed      88.103614    87.843333  
launch_angle      12.909639    11.556667  
spin_rate       2199.084337  2201.983333  
velocity          87.900000    87.506667  
whiffs            96.891566    86.900000  
WAR                0.445783     0.343333  
ERA                4.699759     4.519000  
K/9                8.017711     8.443500  
BB/9               3.802289     3.654333  
WHIP               1.411928     1.376333  
FIP                4.359277     4.356667  
SIERA              4.265301     3.963833  
Hard%              0.306723     0.303067  
Soft%              0.166735     0.166617

## Top - spin_rate
## Mid - BA, ISO, slg, woba,hits, whiffs, WAR, ERA, FIP
## Low - babip, SIERA, Hard%, Soft%

## Low X axis pitchers are best at poor contact and middle x axis pitchers are best
## As long as the pitcher pitches below the 0 in the axis then it should be a decent pitcher to target